### DISCUSSION TITLE
Print arbitrary files directly

### DISCUSSION THREAD

**Discussion by sajadshouli** [NativePHP/laravel#261] [Ideas]
I have some pdf files which must printed directly after user order food in restaurant it means this process must be automatically.  We use direct print google chrome extension.  We pass pdf url and printer name to the direct print functions and the extension do the print

Other ways is write a asp.net application and run it on windows to do that since in asp.net we access to printers

Is there a way to do that with native php framwork? like create a desktop application with nativephp  and install it on windows with xampp and pass the pdf url and printer name to a route to print pdf

**Comment by digitall-it** [comment]
check out
- printnode
- qz tray

**Comment by basst85** [comment]
@sajadshouli You can now print directly to PDF with ``System::printToPDF('<b>Your HTML code</b>')``. It returns a base64 encoded PDF file, which you can save or print.

**Comment by digitall-it** [comment]
@basst85 it is wonderful that you can print to a pdf, but can you print to a printer?

**Comment by basst85** [comment]
@digitall-it You can also print directly to a printer: ``System::print('<b>Your HTML code</b>', 'printername');``. 
But this function only supports HTML, no a PDF-file. And the printer(name) must be known on the system. 

Maybe it's possible to send a generated PDF to an API for direct/cloud printing?

**Reply by digitall-it** [reply]
Qz tray does that and supports various types of hardware, it works by having a server running in the system tray and has a JavaScript interface for the client. You can print PDFs, you will need to solve some trust issues with certificates (that means that the customer needs to do some hoops to make everything work). 

Printnode even allows to send prints to standalone printers across the network/internet.

**Comment by simonhamp** [comment]
I wasn't aware of this before, but it seems that each platform has a way to print from the command line. On macOS and Linux, there's a built-in tool called `lp` and on Windows it seems you can use [PowerShell to initiate a print job](https://superuser.com/questions/1613038/how-do-you-print-any-printable-file-in-windows-10-by-command-line/1616067#1616067).

You may be able to use these commands directly from your PHP code via `shell_exec()`, but you'll need some logic to check which platform you're on before running the relevant command (i.e. check the value of the `PHP_OS_FAMILY` constant)

I expect at some point we'll have a neat cross-platform wrapper for this behaviour, if someone wants to PR that...
